.emacs.d/config.org
#+title: My Emacs Configuration
#+author: Jason Kenyon
#+property: header-args :tangle init.el


* Font and basic settings
#+begin_src emacs-lisp

;emacs auto-generated settings moved to separate file
(setq custom-file "~/.emacs.d/custom.el")
(load custom-file)
;;needed for fullscreen in qtile
(setq frame-resize-pixelwise t)


;;appearance
(blink-cursor-mode 0)

(set-face-attribute 'default nil :family "SourceCodePro" :height 180)
;;(load-theme 'deeper-blue)
(global-display-line-numbers-mode t)
(setq display-line-numbers-type 'relative)
(setq inhibit-startup-message t)
(scroll-bar-mode -1)
(tool-bar-mode -1)
(tooltip-mode -1)
;;(setq visible-bell t)
(menu-bar-mode -1)


#+end_src

* Package repository initialization
#+begin_src emacs-lisp
;;packages
(require 'package)
(setq package-archives '(("melpa" . "https://melpa.org/packages/")
                                ("elpa" . "https://elpa.gnu.org/packages/")))
(package-initialize)

(unless package-archive-contents
  (package-refresh-contents))

(unless (package-installed-p 'use-package)
  (package-install 'use-package))

(require 'use-package)
(setq use-package-always-ensure t)
#+end_src
* M-X menu
** which key
#+begin_src emacs-lisp
(use-package which-key
  :init
  (setq which-key-side-window-location 'bottom
        which-key-sort-order #'which-key-key-order-alpha
        which-key-sort-uppercase-first nil
        which-key-add-column-padding 1
        which-key-max-display-columns nil
        which-key-min-display-lines 6
        which-key-side-window-slot -10
        which-key-side-window-max-height 0.25
        which-key-idle-delay 0.8
        which-key-max-description-length 25
        which-key-allow-imprecise-window-fit t
        which-key-separator " → " ))
(which-key-mode)

(use-package doom-themes
  :config
  ;; Global settings (defaults)
  (setq doom-themes-enable-bold t    ; if nil, bold is universally disabled
        doom-themes-enable-italic t) ; if nil, italics is universally disabled
  (load-theme 'doom-one t)

  ;; Enable flashing mode-line on errors
  ;;(doom-themes-visual-bell-config)
  ;; Enable custom neotree theme (all-the-icons must be installed!)
  (doom-themes-neotree-config)
  ;; or for treemacs users
  (setq doom-themes-treemacs-theme "doom-atom") ; use "doom-colors" for less minimal icon theme
  (doom-themes-treemacs-config)
  ;; Corrects (and improves) org-mode's native fontification.
  (doom-themes-org-config))
#+end_src
** Ivy
#+begin_src emacs-lisp
(use-package ivy
  :diminish
  :config
  (ivy-mode 1))
  #+end_src

** Doom modeline and preqs
#+begin_src emacs-lisp
(use-package compat)
(use-package all-the-icons)
(use-package doom-modeline)
(doom-modeline-mode 1)



#+end_src

* Keybinding
** Evilmode setup
General makes keybinding with leader keys easy
Evil collection has additional vi-like bindings that standard evil mode lacks
Splitting is defined properly
#+begin_src emacs-lisp
(use-package evil
  :init      ;; tweak evil's configuration before loading it
  (setq evil-want-keybinding nil)
  (setq evil-vsplit-window-right t)
  (setq evil-split-window-below t)
  (evil-mode))
(use-package evil-collection
  :after evil
  :config
  (setq evil-collection-mode-list '(dashboard dired ibuffer mu4e))
  (evil-collection-init))

(use-package general
  :config
  (general-evil-setup t))
#+end_src
** Buffer controls
#+begin_src emacs-lisp
(nvmap :prefix "SPC"
       "b b"   '(ibuffer :which-key "Ibuffer")
       "b c"   '(clone-indirect-buffer-other-window :which-key "Clone indirect buffer other window")
       "b k"   '(kill-current-buffer :which-key "Kill current buffer")
       "b n"   '(next-buffer :which-key "Next buffer")
       "b p"   '(previous-buffer :which-key "Previous buffer")
       "b B"   '(ibuffer-list-buffers :which-key "Ibuffer list buffers")
       "b K"   '(kill-buffer :which-key "Kill buffer"))
#+end_src
** Auxiliary Bindings
#+begin_src emacs-lisp 
(general-define-key
 :keymaps '(insert)
 "jk" 'evil-normal-state)
#+end_src
** File controls
#+begin_src emacs-lisp
(use-package all-the-icons-dired)
(nvmap :prefix "SPC"
       "."     '(find-file :which-key "Find file")
       "f r"   '(counsel-recentf :which-key "Recent files")
       "f s"   '(save-buffer :which-key "Save file")
       "f u"   '(sudo-edit-find-file :which-key "Sudo find file")
       "f C"   '(copy-file :which-key "Copy file")
       "f D"   '(delete-file :which-key "Delete file")
       "f R"   '(rename-file :which-key "Rename file")
       "f S"   '(write-file :which-key "Save file as...")
       "f U"   '(sudo-edit :which-key "Sudo edit file"))
#+end_src
       
** zoom
#+begin_src emacs-lisp
(global-set-key (kbd "C-=") 'text-scale-increase)
(global-set-key (kbd "C--") 'text-scale-decrease)
#+end_src
* Org
** Org and roam general settings
`org-indent-mode` makes headlines and source blocks automatically indent correctly
`visual-line-mode` automatically breaks lines in buffer while viewing but doesn't actually insert a newline character
#+begin_src emacs-lisp
    ;; org stuff
    (add-hook 'org-mode-hook 'visual-line-mode)
    (add-hook 'org-mode-hook 'org-indent-mode)
    (use-package org-roam)
    (setq org-directory "~/projects/org/")
    (setq org-roam-directory "~/projects/org/org-roam/")
    (setq org-agenda-files '("~/projects/org/org-roam/daily/" "~/projects/org/org-roam"))
    (org-roam-db-autosync-mode)
    (setq org-clock-persist 'history)
    (org-clock-persistence-insinuate)


    (setq org-roam-dailies-directory "daily/")
    (setq org-roam-dailies-capture-templates
        '(("d" "default" entry
            "* %?"
            :target (file+head "%<%Y-%m-%d>.org"
                                "#+title: %<%Y-%m-%d>\n"))))
(use-package org-bullets)
(setq org-ellipsis " ▼ ")
(add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))
#+end_src
** bindings
#+begin_src emacs-lisp

(nvmap :keymaps 'org-mode-map
"TAB" 'org-cycle)


    (nvmap :keymaps 'override :prefix "SPC"
           "m *"   '(org-ctrl-c-star :which-key "Org-ctrl-c-star")
           "m +"   '(org-ctrl-c-minus :which-key "Org-ctrl-c-minus")
           "m e"   '(org-export-dispatch :which-key "Org export dispatch")
           "m f"   '(org-footnote-new :which-key "Org footnote new")
           "m h"   '(org-toggle-heading :which-key "Org toggle heading")
           "m i"   '(org-toggle-item :which-key "Org toggle item")
           "m n"   '(org-store-link :which-key "Org store link")
           "m o"   '(org-set-property :which-key "Org set property")
           "m t"   '(org-todo :which-key "Org todo")
           "m x"   '(org-toggle-checkbox :which-key "Org toggle checkbox")
           "m B"   '(org-babel-tangle :which-key "Org babel tangle")
           "m I"   '(org-toggle-inline-images :which-key "Org toggle inline imager")
           "m T"   '(org-todo-list :which-key "Org todo list")
           "o a"   '(org-agenda :which-key "Org agenda")
           )
#+end_src
** Keywords and source block config
org-tempo allows for `<s` like expansions
#+begin_src emacs-lisp
  (setq org-todo-keywords        ; This overwrites the default Doom org-todo-keywords
          '((sequence
             "TODO(t)"           ; A task that is ready to be tackled
             "PROJ(p)"           ; A project that contains other tasks
             "WAIT(w)"           ; Something is holding up this task
             "DONE(d)"           ; Task has been completed
              )))  ; Task has been cancelled
(require 'org-tempo)
(setq org-src-fontify-natively t
  org-src-tab-acts-natively t
  org-confirm-babel-evaluate nil
  org-edit-src-content-indentation 0)  

#+end_src

* LSP and Completion
#+begin_src emacs-lisp
(use-package lsp-mode
:init
(setq lsp-keymap-prefix "C-c l")
:hook
(lsp-mode . lsp-enable-which-key-integration)
(lsp-mode . python-mode)
(lsp-mode . latex-mode))
(setq lsp-pyls-server-command "$HOME/.local/bin/pyls")
(use-package lsp-ui :commands lsp-ui-mode)
(use-package lsp-ivy :commands lsp-ivy-workspace-symbol)
#+end_src

* mail
#+begin_src emacs-lisp
(use-package mu4e
  :ensure nil
  :load-path "/usr/share/emacs/site-lisp/mu4e/"
  :config
  (setq mu4e-change-filenames-when-moving t)
  (setq mu4e-update-interval (* 10 60))
  (setq mu4e-get-mail-command "mbsync -a")
  (setq mu4e-maildir "~/Mail")
  (setq mu4e-drafts-folder "/[Gmail/.Drafts")
  (setq mu4e-sent-folder "/[Gmail/.Sent Mail")
  (setq mu4e-refile-folder "/[Gmail/.All Mail")
  (setq mu4e-trash-folder "/[Gmail/.Trash")

  (setq mu4e-maildir-shortcuts
   '(("/Inbox" . ?i)
    ("/[Gmail].Sent Mail" . ?s)
    ("/[Gmail].All Mail" . ?a)
    ("/[Gmail].Drafts" . ?d)
    ("/[Gmail].Trash" . ?t))))



#+end_src


.config/doom/config.org
#+title: Doom Emacs Configuration
#+author: Jason Kenyon
#+property: header-args :tangle config.el


* Font and Modeline
#+begin_src emacs-lisp

;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!


;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets. It is optional.
(setq user-full-name "Jason Kenyon"
      user-mail-address "jason0kenyon@gmail.com")

;; Doom exposes five (optional) variables for controlling fonts in Doom:
;;
;; - `doom-font' -- the primary font to use
;; - `doom-variable-pitch-font' -- a non-monospace font (where applicable)
;; - `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;; - `doom-unicode-font' -- for unicode glyphs
;; - `doom-serif-font' -- for the `fixed-pitch-serif' face
;;
;; See 'C-h v doom-font' for documentation and more examples of what they
;; accept. For example:
;;
;;(setq doom-font (font-spec :family "Fira Code" :size 12 :weight 'semi-light)
;;      doom-variable-pitch-font (font-spec :family "Fira Sans" :size 13))
;;
;; If you or Emacs can't find your font, use 'M-x describe-font' to look them
;; up, `M-x eval-region' to execute elisp code, and 'M-x doom/reload-font' to
;; refresh your font settings. If Emacs still can't find your font, it likely
;; wasn't installed correctly. Font issues are rarely Doom issues!

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
(setq doom-variable-pitch-font (font-spec :family "SourceSansPro" :size 36 ))
(setq doom-big-font (font-spec :family "SourceCodePro" :weight 'semibold :size 46 ))
(setq doom-font (font-spec :family "SourceCodePro" :size 36 :weight 'semibold ))

(setq doom-theme 'doom-dark+)

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type 'relative)
(setq doom-modeline-height 70)

#+end_src

* Be Warned
#+begin_src emacs-lisp
;; Whenever you reconfigure a package, make sure to wrap your config in an
;; `after!' block, otherwise Doom's defaults may override your settings. E.g.
;;
;;   (after! PACKAGE
;;     (setq x y))
;;
;; The exceptions to this rule:
;;
;;   - Setting file/directory variables (like `org-directory')
;;   - Setting variables which explicitly tell you to set them before their
;;     package is loaded (see 'C-h v VARIABLE' to look up their documentation).
;;   - Setting doom variables (which start with 'doom-' or '+').
;;
;; Here are some additional functions/macros that will help you configure Doom.
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;; Alternatively, use `C-h o' to look up a symbol (functions, variables, faces,
;; etc).
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.


;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
#+end_src
* Org Mode and Roam
Setting org directory: this is where org-mode searches for keywords and scheduling, e.g,  TODO entries.
Similarly, I set the daily directory for Roam with a simple template that is generated every time I open a daily org document via Roam.
#+begin_src emacs-lisp

(use-package! org-roam-ui
    :after org-roam ;; or :after org
;;         normally we'd recommend hooking orui after org-roam, but since org-roam does not have
;;         a hookable mode anymore, you're advised to pick something yourself
;;         if you don't care about startup time, use
;;  :hook (after-init . org-roam-ui-mode)
    :config
    (setq org-roam-ui-sync-theme t
          org-roam-ui-follow t
          org-roam-ui-update-on-save t
          org-roam-ui-open-on-start t))
(after! org

        (setq org-directory "~/projects/org/")
        (setq org-roam-directory "~/projects/org/org-roam/")
        (setq org-agenda-files '("~/projects/org/org-roam/daily/" "~/projects/org/org-roam"))
                )

        (setq org-roam-dailies-directory "daily/")
        (setq org-roam-dailies-capture-templates
            '(("d" "default" entry
                "* %?"
                :target (file+head "%<%Y-%m-%d>.org"
                                    "#+title: %<%Y-%m-%d>\n"))))
#+end_src
* Mu4e Email Client
Configured using Doom Emacs macro `set-email-account`
Shortcuts are visble from mu4e menu and allow for easy access to specific inboxes
`mu4e-x-folder` declarations have paths relative to the maildir declared using mu initially, i.e., with `mu init`
#+begin_src emacs-lisp
(set-email-account! "jason0kenyon"
  '((mu4e-sent-folder       . "/jason0kenyon/Sent Mail")
    (mu4e-drafts-folder     . "/jason0kenyon/Drafts")
    (mu4e-trash-folder      . "/jason0kenyon/Trash")
    (mu4e-refile-folder     . "/jason0kenyon/All Mail")
    (smtpmail-smtp-user     . "jason0kenyon@gmail.com")
    (user-mail-address      . "jason0kenyon@gmail.com")    ;; only needed for mu < 1.4
    (mu4e-compose-signature . "---\nJason Kenyon"))
  t)


(after! mu4e
(setq mu4e-maildir-shortcuts
    '((:maildir "/Inbox"    :key ?i)
      (:maildir "/[Gmail]/Sent Mail" :key ?s)
      (:maildir "/[Gmail]/Trash"     :key ?t)
      (:maildir "/[Gmail]/Drafts"    :key ?d)
      (:maildir "/[Gmail]/All Mail"  :key ?a)))

  )


#+end_src


.config/doom/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!


;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets. It is optional.
(setq user-full-name "Jason Kenyon"
      user-mail-address "jason0kenyon@gmail.com")

;; Doom exposes five (optional) variables for controlling fonts in Doom:
;;
;; - `doom-font' -- the primary font to use
;; - `doom-variable-pitch-font' -- a non-monospace font (where applicable)
;; - `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;; - `doom-unicode-font' -- for unicode glyphs
;; - `doom-serif-font' -- for the `fixed-pitch-serif' face
;;
;; See 'C-h v doom-font' for documentation and more examples of what they
;; accept. For example:
;;
;;(setq doom-font (font-spec :family "Fira Code" :size 12 :weight 'semi-light)
;;      doom-variable-pitch-font (font-spec :family "Fira Sans" :size 13))
;;
;; If you or Emacs can't find your font, use 'M-x describe-font' to look them
;; up, `M-x eval-region' to execute elisp code, and 'M-x doom/reload-font' to
;; refresh your font settings. If Emacs still can't find your font, it likely
;; wasn't installed correctly. Font issues are rarely Doom issues!

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
(setq doom-variable-pitch-font (font-spec :family "SourceSansPro" :size 36 ))
(setq doom-big-font (font-spec :family "SourceCodePro" :weight 'semibold :size 46 ))
(setq doom-font (font-spec :family "SourceCodePro" :size 36 :weight 'semibold ))

(setq doom-theme 'doom-dark+)

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type 'relative)
(setq doom-modeline-height 70)

;; Whenever you reconfigure a package, make sure to wrap your config in an
;; `after!' block, otherwise Doom's defaults may override your settings. E.g.
;;
;;   (after! PACKAGE
;;     (setq x y))
;;
;; The exceptions to this rule:
;;
;;   - Setting file/directory variables (like `org-directory')
;;   - Setting variables which explicitly tell you to set them before their
;;     package is loaded (see 'C-h v VARIABLE' to look up their documentation).
;;   - Setting doom variables (which start with 'doom-' or '+').
;;
;; Here are some additional functions/macros that will help you configure Doom.
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;; Alternatively, use `C-h o' to look up a symbol (functions, variables, faces,
;; etc).
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.


;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!

(use-package! org-roam-ui
    :after org-roam ;; or :after org
;;         normally we'd recommend hooking orui after org-roam, but since org-roam does not have
;;         a hookable mode anymore, you're advised to pick something yourself
;;         if you don't care about startup time, use
;;  :hook (after-init . org-roam-ui-mode)
    :config
    (setq org-roam-ui-sync-theme t
          org-roam-ui-follow t
          org-roam-ui-update-on-save t
          org-roam-ui-open-on-start t))
(after! org

        (setq org-directory "~/projects/org/")
        (setq org-roam-directory "~/projects/org/org-roam/")
        (setq org-agenda-files '("~/projects/org/org-roam/daily/" "~/projects/org/org-roam"))
                )

        (setq org-roam-dailies-directory "daily/")
        (setq org-roam-dailies-capture-templates
            '(("d" "default" entry
                "* %?"
                :target (file+head "%<%Y-%m-%d>.org"
                                    "#+title: %<%Y-%m-%d>\n"))))

(set-email-account! "jason0kenyon"
  '((mu4e-sent-folder       . "/jason0kenyon/Sent Mail")
    (mu4e-drafts-folder     . "/jason0kenyon/Drafts")
    (mu4e-trash-folder      . "/jason0kenyon/Trash")
    (mu4e-refile-folder     . "/jason0kenyon/All Mail")
    (smtpmail-smtp-user     . "jason0kenyon@gmail.com")
    (user-mail-address      . "jason0kenyon@gmail.com")    ;; only needed for mu < 1.4
    (mu4e-compose-signature . "---\nJason Kenyon"))
  t)


(after! mu4e
(setq mu4e-maildir-shortcuts
    '((:maildir "/Inbox"    :key ?i)
      (:maildir "/[Gmail]/Sent Mail" :key ?s)
      (:maildir "/[Gmail]/Trash"     :key ?t)
      (:maildir "/[Gmail]/Drafts"    :key ?d)
      (:maildir "/[Gmail]/All Mail"  :key ?a)))

  )
